### PR TITLE
Huber Loss Function

### PR DESCRIPTION
**Pull request by Hinal-Srivastava**
**Huber Loss Function**
Huber loss approaches MSE when 𝛿 ~ 0, i.e for smaller values and MAE when 𝛿 ~ ∞, i.e for larger values.
Viyet
**References:** 
1. [Documentation]( https://towardsdatascience.com/understanding-the-3-most-common-loss-functions-for-machine-learning-regression-23e0ef3e14d3) on Huber Loss Function 
2. [Video ](https://www.youtube.com/watch?v=Kj1TFLexfw8)on Huber Loss Function 

Reference Issue #21

### CODE REVIEW

File: MLlib/loss_func.py [hunk 1]
```diff
@@ -26,3 +26,17 @@ def derivative(X, Y, W):
         M = X.shape[0]
         H = sigmoid(np.dot(X, W).T)
         return (1/M)*(np.dot(X.T, (H-Y).T))
+
+class Huber_Loss_Function():
+    @staticmethod
+    def loss(m, b, X, Y, delta):
```
Comment (agrawalshubham01) [inline MLlib/loss_func.py:32]:
```suggestion
    def loss(m, b, X, Y, delta= 0.00001):
```

Rest LGTM :-)

**Review by agrawalshubham01 — DISMISSED** [review]

**Comment by agrawalshubham01** [comment]
@Hinal-Srivastava 
https://github.com/RoboticsClubIITJ/ML-DL-implementation/blob/d73e1460762635cff3892e8f168dbb64eadba5ca/MLlib/loss_func.py#L5-L15
- Here only 3 parameters are taken i.e instead of `W*X+B` we are just using W*X reason being we have added bias row/column. Kindly Correct this in your code.
- Also you have in `loss` only loss is calculated and is independent of iteration (Review the above code), And add function `derivative` which calculates derivative of the function given

@rohansingh9001 any other point need to be added?

**Comment by Hinal-Srivastava** [comment]
> @Hinal-Srivastava
> https://github.com/RoboticsClubIITJ/ML-DL-implementation/blob/d73e1460762635cff3892e8f168dbb64eadba5ca/MLlib/loss_func.py#L5-L15
> 
> * Here only 3 parameters are taken i.e instead of `W*X+B` we are just using W*X reason being we have added bias row/column. Kindly Correct this in your code.
> * Also you have in `loss` only loss is calculated and is independent of iteration (Review the above code), And add function `derivative` which calculates derivative of the function given
> 
> @rohansingh9001 any other point need to be added?

Yes I'll change these ASAP. Sorry for the late reply

File: MLlib/loss_func.py [hunk 2]
```diff
@@ -26,3 +26,26 @@ def derivative(X, Y, W):
...
+        B = sigmoid(np.dot(X, W).T)
+        N = len(X)
+        for i in range(N):            
+            if abs(Y[i] - M*X[i] - B) <= delta: #Quadratic Derivative for <=delta
+                M = -X[i] * (Y[i] - (M*X[i] + B))
+                B = - (Y[i] - (M*X[i] + B))
+            else:
+                M = delta * X[i] * ((M*X[i] + B) - Y[i]) / abs((M*X[i] + B) - Y[i]) #Linear Derivative for >delta
+                B = delta * ((M*X[i] + B) - Y[i]) / abs((M*X[i] + B) - Y[i])
+        return M, B
```
Comment (agrawalshubham01) [inline MLlib/loss_func.py:43]: Can't this be done like

```
M= X.shape[0]
mse = 0.5*(np.dot(X, W).T - Y)**2
mae = delta * (np.abs(np.dot(X, W).T - Y) - 0.5 * delta**2)
return np.sum(np.where(np.abs(np.dot(X, W).T - Y) <= delta, mse, mae)
```

**Comment by rohansingh9001** [comment]
There are merge errors, build errors and no updates for a long time. Closing this PR for now. The author can open a new PR when it is done.

**Comment by TarunTomar122** [comment]
Closed Due to Inactivity!

**Pull request closed without merging**